src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const transformedState = structuredClone(state);
+  const stateHistory = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    if (actions[i].type === 'addProperties') {
+      for (const key in actions[i].extraData) {
+        transformedState[key] = actions[i].extraData[key];
+      }
+    } else if (actions[i].type === 'removeProperties') {
+      for (const key of actions[i].keysToRemove) {
+        delete transformedState[key];
+      }
+    } else {
+      for (const key in transformedState) {
+        delete transformedState[key];
+      }
```

Comment:
The current approach is repeating the action of pushing the `transformedState` into `stateHistory` after every case. This should be done once after the switch statement to adhere to the DRY principle as per checklist item 4.